### ISSUE TITLE
data-default-instances-old-locale build fails on Windows with GHC 7.10.1 and Cabal 1.22.2.0

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/cabal#2502]
I'm not sure what's going on, but for some reason, `data-default-instances-old-locale` and `data-default-instances-containers` will not build if I try to install them directly from Hackage. Here's the output from trying to install `data-default-instances-old-locale`:

```
C:\...\ghc-7.10.1\mingw\bin\ar.exe: dist\build\libHSdata-default-instances-old-locale-0.0.1-6jcjjaR25tK4x3nJhHHjFM.a-6440\libHSdata-default-instances-old-locale-0.0.1-6jcjjaR25tK4x3nJhHHjFM.a: No such file or directory
```

Here's a [gist](https://gist.github.com/RyanGlScott/78a06d74f35f0b271d64) of the full `cabal install -j1 -v3` output. I am using 64-bit Windows 7 with GHC 7.10.1, `Cabal-1.22.0.0` and `cabal-install-1.22.0.0`, and I can reproduce this with `cmd.exe` or PowerShell.

`data-default-instances-old-locale` and `data-default-instances-containers` are the only two libraries that give me this error. Because of this, I'd hazard a guess that it's due to the length of the names of the packages themselves that are causing problems, since the temporary directories that `cmd.exe` creates might exceed some Windows path length constant somewhere. Unfortunately, I can't find any other Hackage packages that are as long as those two that are compatible with GHC 7.10.1, so it's hard to say.

It should be noted that this error does not happen if you try installing from Cygwin or MSYS. It also does not happen if you download the libraries with `cabal get` and install them directly (no matter what shell is used).

**Comment by tomasmcz** [comment]
The same thing happened to me in Cygwin, but in my case the error only manifested itself when I let Cabal use multiple jobs, it went away with `-j1`.

**Comment by solatis** [comment]
Just confirming that I am also having this same error with minghc 7.10 under windows 8.1

**Comment by Twinside** [comment]
So I found another package getting problem (from a duplicate issue I opened): vector-binary-instances

**Comment by vagarenko** [comment]
blaze-builder-enumerator-0.2.1.0 also fails with message
```
E:\Program Files (x86)\MinGHC-7.10.1\ghc-7.10.1\mingw\bin\ar.exe: dist/dist-sand
box-e2f01390\build\libHSblaze-builder-enumerator-0.2.1.0-ItlkLd1FDlt1WK6wQPx1K1.
a-11148\libHSblaze-builder-enumerator-0.2.1.0-ItlkLd1FDlt1WK6wQPx1K1.a: No such
file or directory
```

**Comment by vagarenko** [comment]
and MonadCatchIO-transformers-0.3.1.3 too

**Comment by dmcclean** [comment]
This happens to me too on minghc7.10 on windows 7 professional.
It happens for data-default-instances-base, data-default-instances-containers, data-default-instances-dlist, and data-default-instances-old-locale.
The suggested fixes of using `-j1` does not work for me, but separating the get and install phases as suggested in #2525 does work.

**Comment by mantkiew** [comment]
@dmcclean, cabal install directly fails on Win 8.1 minghc7.10.1 but the trick with separating get and install phases works as well.

**Comment by gzh** [comment]
Experienced same issue, Windows 7 x86_64, ghc 7.10. Neither -j1 nor separate get & install helped; that was cabal-install 1.22.2.0 built against cabal library 1.22.2.0. I decided to rollback to cabal-install 1.22.0.1 and cabal library 1.22.1.1; with this version I managed to build required libraries with -j1 switch (however build still failed with -jN, N>1).

**Comment by snoyberg** [comment]
Discovered during my testing as well.

**Comment by snoyberg** [comment]
This has been open for 1.5 months, and I'm getting more user complaints. Is there any ETA on a fix for this? I'm also quite concerned that the milestone is set for 1.24, when I would have hoped that this would warrant an emergency hotfix given the severity. I'm considering telling Windows users that they need to avoid GHC 7.10 entirely because of this.

**Comment by 23Skidoo** [comment]
@ezyang, can this be related to the package key changes?

**Comment by melted** [comment]
I see the same for data-default-* when building idris in a sandbox. My suspicion is that the new file and directory names that are very lavish with chars are sneaking up against the 260 char path limit on windows.

**Comment by dmcclean** [comment]
I had that suspicion too. I tried to test it by setting up a sandbox in a one-letter directory at the root of a drive to see if it would install there, and it didn't change the error, so I figured I was barking up the wrong tree. But it is still possible, that test doesn't really rule it out.

**Comment by hsenag** [comment]
I looked at this with `procmon`, and `ar` is trying to work with
`dist\build\libHSdata-default-instances-old-locale-0.0.1-6jcjjaR25tK4x3nJhHHjFM.a-11924\libHSdata-default-instances-old-locale-0.0.1-6jcjjaR25tK4x3nJhHHjFM.a`
in a working directory of
`C:\Users\Ganesh\AppData\Local\Temp\data-default-instances-old-locale-0.0.1-11924\data-default-instances-old-locale-0.0.1`

The absolute path for this file would be 278 characters.

When I set the `TEMP` and `TMP` environment variables to something short, like `c:\temp\`, the problem goes away. I guess that would be a good workaround for the problem for now. Hopefully it's not strictly necessary to repeat the package name 4 times in the directory structure :-)

**Comment by ezyang** [comment]
Hey guys: This problem was tickled by #2437. It would be relatively easy to revert that fix, but it sounds like people like descriptive names. Probably Cabal can be fixed to use a different TMP path, but I think there are some constraints on what filenames ar expects / deterministic names to make sure build products stay stable.

**Comment by NightRa** [comment]
Same problem here.

**Comment by ezyang** [comment]
Today I did some source code snooping to find where the temporary directories are being made. The operant function is with `withTempDirectory`.

The working directory is created by `installLocalTarballPackage` in `Distribution.Client.Install`, which uses `display pkgid` as a template to make a temporary directory (component 1), and then unpacks into the temporary directory into the name of the pkgid (component 2); we can change this part to only use the pkgid in the temporary directory and don't extract into a nested folder; however, none of these folder names are strictly necessary.

Component 3 comes from Cabal.Distribution.SimpleProgram.Ar use  in `createArLibArchive` which uses `targetName` as the template of the temp dir. Component (4) is strictly necessary since we need it to identify the libraries.

Here's a simple patch that will probably fix the problem for people:

{{{
diff --git a/Cabal/Distribution/Simple/Program/Ar.hs b/Cabal/Distribution/Simple/Program/Ar.hs
index 32d3a70..6d1b3bc 100644
--- a/Cabal/Distribution/Simple/Program/Ar.hs
+++ b/Cabal/Distribution/Simple/Program/Ar.hs
@@ -48,7 +48,7 @@ createArLibArchive verbosity lbi targetPath files = do
   (ar, _) <- requireProgram verbosity arProgram progConf
 
   let (targetDir, targetName) = splitFileName targetPath
-  withTempDirectory verbosity targetDir targetName $ \ tmpDir -> do
+  withTempDirectory verbosity targetDir "objs" $ \ tmpDir -> do
   let tmpPath = tmpDir </> targetName
 
   -- The args to use with "ar" are actually rather subtle and system-dependent.
nstall.hst a/cabal-install/Distribution/Client/Install.hs b/cabal-install/Distribution/Client/ 
index 5ae90af..13d9e8c 100644   
--- a/cabal-install/Distribution/Client/Install.hs
+++ b/cabal-install/Distribution/Client/Install.hs
@@ -1269,8 +1269,8 @@ installLocalTarballPackage verbosity jobLimit pkgid
   tmp <- getTemporaryDirectory
   withTempDirectory verbosity tmp (display pkgid) $ \tmpDirPath ->
     onFailure UnpackFailed $ do
-      let relUnpackedPath = display pkgid
-          absUnpackedPath = tmpDirPath </> relUnpackedPath
+      let relUnpackedPath = "."
+          absUnpackedPath = tmpDirPath
           descFilePath = absUnpackedPath
                      </> display (packageName pkgid) <.> "cabal"
       withJobLimit jobLimit $ do
}}}

A more general fix might be to make `withTempDirectory` truncate the template if it's too long.

Relatedly, we might consider compressing the names of the caches, which currently take the form `.cabal/packages/hackage.haskell.org/packagename/0.0.0/packagename-0.0.0.tar.gz`. Relevant functions are `packageFile` and `packageDir` in FetchUtils.

**Comment by ezyang** [comment]
(BTW, I'll block for a little bit on discussion on how people would to fix this, if no one pipes up I'll just commit the patch above.)

**Comment by wismill** [comment]
Have you considered to prefix the path with `\\?\` to [lift the restriction of length](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx#maxpath)?

**Comment by 23Skidoo** [comment]
@rthomas, can you please cut a new Cabal point release from the 1.22 branch? We want this fix in GHC 7.10.2.

**Comment by rthomas** [comment]
Yep will do - sorry about the delay.
